### PR TITLE
gpath: add recursive support

### LINKED ISSUE #2634: Add recursive support for `GPath`
Currently, `tf.io.gfile.glob` does not support recursive: https://github.com/tensorflow/tensorflow/issues/39565

This make `GPath().glob('**/x.txt')` inconsistent with python `glob`.

It would be nice to add recursive support. This would be done by a combinaison of `.iterdir()`, `.isdir()` (to recurse) and [`fnmatch.fnmatch()`](https://docs.python.org/3/library/fnmatch.html#fnmatch.fnmatch) to check that the file match the pattern.

### PR DESCRIPTION
**Pull request by iamsdas**
Fix #2634
Note: This is currently supposed be used something like `Gpath('src').rglob('**/*.py')`. currently it will fail if there is any absolute path before `**` in the pattern.

### CODE REVIEW

File: tensorflow_datasets/core/utils/gpath.py [hunk 1]
```diff
@@ -89,6 +89,16 @@ def resolve(self: _P, strict: bool = False) -> _P:
...
+    """
+    Yielding all matching files (of any kind) recursively.
+    """
+    yield from self.glob(pattern)
+    if '**' in pattern:
+      for f in self.iterdir():
+        if f.is_dir():
+          yield from f.rglob(pattern)
+
   def glob(self: _P, pattern: str) -> Iterator[_P]:
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/utils/gpath.py:102]: To be compatible with `pathlib.Path` API, `Path().glob('**/*.jpeg')` should also be recursive. See https://docs.python.org/3/library/pathlib.html#pathlib.Path.glob for doc.

**Review by Conchylicultor — CHANGES_REQUESTED** [review]
Thank you for starting this. Could you check add tests to make sure this is working ?

File: tensorflow_datasets/core/utils/gpath.py (hunk 1 above)
Comment (iamsdas) [reply tensorflow_datasets/core/utils/gpath.py:102]: I'm not sure I understand what you mean to say over here. Currently, `Gpath().glob('**/*.jpeg)` does work as intended. At the beginning of the function, it uses `self.glob()` to find matching patterns in the given path (which are recursively called at the bottom).

File: tensorflow_datasets/core/utils/gpath.py [hunk 2]
```diff
@@ -89,6 +89,16 @@ def resolve(self: _P, strict: bool = False) -> _P:
     """Returns the abolute path."""
     return self._new(posixpath.abspath(self._path_str))
 
+  def rglob(self: _P, pattern: str) -> Iterator[_P]:
```
[the remark this answers is not in the document]
Comment (iamsdas) [reply tensorflow_datasets/core/utils/gpath.py:92]: Although these two examples will work, something like `Gpath('/tmp').glob('a/**/*.txt)` will not (patterns with absolute path before **). This is a WIP. Im currently thinking of a hacky way to get around this by splitting the pattern and adding the part before `**` to the path and removing it from the pattern. Is this ok? I'll add the tests after doing this.

File: tensorflow_datasets/core/utils/gpath.py (hunk 1 above)
Comment (Conchylicultor) [reply tensorflow_datasets/core/utils/gpath.py:102]: You could have a look at the `glob.iglob` implementation to see how they are doing it. There might be smarter way to do this

Comment (iamsdas) [reply tensorflow_datasets/core/utils/gpath.py:102]: Is there any reason to not directly use something like `pathlib.Path().rglob(pattern)` and pass it to a for loop to make it our desired type instead? The implementation of `glob.iglob` can get a bit complecated as we are trying to use `Gpath(path).rglob(pattern)` which has this exrta `path`, which is not present in `glob.glob()`.

Comment (Conchylicultor) [reply tensorflow_datasets/core/utils/gpath.py:102]: I'm not sure I understand. `pathlib.Path().rglob(pattern)` won't works with GCS right ?
`GPath('gs://some-path/').rglob('*.jpg')`

Comment (iamsdas) [reply tensorflow_datasets/core/utils/gpath.py:102]: Oh got it, I had no idea about that.

Comment (iamsdas) [reply tensorflow_datasets/core/utils/gpath.py:102]: I'll see if i can do a better implementation, until then I'll send a commit which is supposed to work properly in all the cases.

File: tensorflow_datasets/core/utils/gpath.py [hunk 3]
```diff
@@ -18,6 +18,7 @@
 import os
 import pathlib
 import posixpath
+from fnmatch import fnmatch
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/utils/gpath.py:21]: Direct imports are not allowed (use `import fnmatch`, then `fnmatch .fnmatch()`)

File: tensorflow_datasets/core/utils/gpath.py [hunk 4]
```diff
@@ -89,6 +90,22 @@ def resolve(self: _P, strict: bool = False) -> _P:
     """Returns the abolute path."""
     return self._new(posixpath.abspath(self._path_str))
 
+  def iglob(self: _P, pattern: str) -> Iterator[_P]:
```
Comment (Conchylicultor) [inline tensorflow_datasets/core/utils/gpath.py:93]: Rename `iglob` -> `glob`

**Review by Conchylicultor — CHANGES_REQUESTED** [review]
Thank you!. Could you add tests in `gpath_tests.py` ?

File: tensorflow_datasets/core/utils/gpath.py (hunk 3 above)
Comment (iamsdas) [reply tensorflow_datasets/core/utils/gpath.py]: done

**Comment by HoltSpalding** [comment]
Thank You!

**Pull request closed without merging**